### PR TITLE
feat(otel): W3C Trace Context propagation (#4745)

### LINKED ISSUE #4745: feat: OpenTelemetry W3C trace context 전파 도입
## 배경

MASC 멀티 에이전트 간 요청 추적에 trace context가 없어서 크로스 에이전트 디버깅이 어렵다.
GitHub Copilot SDK(2026-04-02)에서 OpenTelemetry W3C trace context 전파를 기본 내장한 것을 확인, 우선순위 상향 근거로 기록.

- 출처: https://github.blog/changelog/2026-04-02-copilot-sdk-in-public-preview/
- GeekNews: https://news.hada.io/topic?id=28149

## 현재 문제

- keeper 간 메시지 전달 시 요청 출처 추적 불가
- broadcast -> claim -> execute 체인에서 어느 에이전트가 병목인지 파악 어려움
- 에러 발생 시 room 내 어떤 에이전트의 어떤 tool 호출에서 실패했는지 로그 산재

## 제안

- MCP 요청/응답에 W3C traceparent 헤더 전파
- MASC broadcast/heartbeat에 trace context 포함
- OCaml OpenTelemetry 라이브러리 조사 (ocaml-opentelemetry 등)

## 제약

- OCaml 생태계의 OpenTelemetry 지원이 제한적일 수 있음 -- 최소한 trace ID 수동 전파부터 시작
- 성능 오버헤드 측정 필요

### PR DESCRIPTION
**Pull request by jeong-sik**
## Summary
- W3C traceparent 파서/생성 모듈 추가 (opentelemetry 0.13 라이브러리 API 활용)
- HTTP 요청에서 traceparent 헤더 추출 -> ambient scope 설정
- Broadcast message/SSE notification에 trace_context 필드 주입
- 27개 단위 테스트 (parsing, roundtrip, propagation, JSON, HTTP headers, edge cases)

## Product impact
MASC 멀티 에이전트 간 요청 추적 기반 마련. 크로스 에이전트 디버깅 시 trace 연결 가능.

## Evidence
- W3C Trace Context 공식 스펙 준수: https://www.w3.org/TR/trace-context/
- opentelemetry 0.13의 Span_ctx.of_w3c_trace_context 활용 (all-zero 검증은 wrapper에서 추가)
- 미래 버전(01) forward-compatibility 테스트 통과 확인

## Review evidence
- GLM-5.1 교차 모델 리뷰 완료 (2026-04-03)
- 7개 지적사항 중 3개 수정 (hot path 최적화, whitespace trim, 복수 헤더 discard)
- 4개 수용 불필요 판정 (tracestate 후속, from_ambient 1회/broadcast, JSON echo 정상, 미래버전 통과)

## Linked issue
Closes #4745

## Design
W3C Trace Context 스펙 준수.

전파 흐름:
```
External HTTP -> traceparent 헤더 -> ambient scope
  -> tool call -> broadcast -> JSON에 trace_context 삽입
    -> SSE -> 수신 에이전트가 JSON에서 추출 가능
```

## Scope
- otel_spans.ml, otel_dispatch_hook.ml, Prometheus 메트릭은 미접촉
- Keeper 메시지 수신 시 trace 복원은 후속 작업
- tracestate 헤더 전파는 후속 작업

## Test plan
- [x] otel_trace_context: 27 tests (parse/generate/propagate/json/headers/edge cases)
- [x] test_types: message roundtrip with trace_context field
- [x] test_http_server_eio: 19 tests passing
- [ ] keeper trace restoration (follow-up)
- [ ] tracestate propagation (follow-up)

### CODE REVIEW

File: lib/http_server_eio.ml [hunk 1]
```diff
@@ -580,12 +580,36 @@ let with_streamable_mcp_request_handler ~request_handler routes =
...
 
 (** Create httpun request handler from router
-    Note: httpun-eio wraps reqd in Gluten.Reqd.t, extract with .reqd field *)
+    Note: httpun-eio wraps reqd in Gluten.Reqd.t, extract with .reqd field
+
+    W3C Trace Context: when MASC_OTEL_ENABLED=true and a valid [traceparent]
+    header is present, the request is dispatched inside an ambient scope
+    carrying the incoming trace_id. Downstream code (tool calls, broadcasts)
+    can read it via [Otel_trace_context.from_ambient()].
+    When disabled or absent, dispatch proceeds without overhead. *)
```
[the remark this answers is not in the document]
Comment (jeong-sik) [reply lib/http_server_eio.ml]: Valid. Comment is misleading — header conversion still happens when OTel enabled. Will correct.
